pkg/cloudprovider/provider/cherryservers/provider.go

```diff
@@ -0,0 +1,429 @@
+	}
+
+	plans, _, err := client.Plans.List(tid)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, plan := range plans {
+		if plan.Name == planName {
+			return &plan, nil
+		}
+	}
+	return nil, nil
+}
+
+func getServerByTag(client *cherrygo.Client, tag string, projectID string) (*cherrygo.Server, error) {
+	servers, _, err := client.Servers.List(projectID)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, server := range servers {
+		if server.Tags[machineUIDTagKey] == tag {
+			srv, _, err := client.Server.List(strconv.Itoa(server.ID))
+			if err != nil {
+				return nil, err
+			}
+			return &srv, nil
+		}
+	}
+
+	return nil, nil
+}
+
+func (p *provider) Validate(spec v1alpha1.MachineSpec) error {
+	c, pc, err := p.getConfig(spec.ProviderSpec)
+	if err != nil {
+		return fmt.Errorf("failed to parse config: %v", err)
+	}
+
+	if c.Token == "" {
+		return errors.New("token is missing")
+	}
+
+	_, err = getNameForOS(pc.OperatingSystem)
+	if err != nil {
+		return fmt.Errorf("invalid/not supported operating system specified %q: %v", pc.OperatingSystem, err)
+	}
+
+	client := getClient(c.Token)
+
+	if c.TeamID != "" {
+		if _, err := getTeamByID(client, c.TeamID); err != nil {
+			return fmt.Errorf("failed to get teamID")
+		}
+	}
+
+	if c.Plan != "" {
+		if _, err := getPlanByName(client, c.Plan, c.TeamID); err != nil {
+			return fmt.Errorf("failed to get plan")
+		}
+	}
+
+	if c.Location != "" {
+		if c.Location != "EU-East-1" && c.Location != "EU-West-1" {
+			return fmt.Errorf("failed to get location")
+		}
+	}
+
+	if c.ProjectID != "" {
+		if _, _, err := client.Project.List(c.ProjectID); err != nil {
+			return fmt.Errorf("failed to get projectID")
+		}
+	}
+
+	return nil
+}
+
+func (p *provider) Create(machine *v1alpha1.Machine, _ *cloudprovidertypes.ProviderData, userdata string) (instance.Instance, error) {
+	c, pc, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Failed to parse MachineSpec, due to %v", err),
+		}
+	}
+
+	client := getClient(c.Token)
+
+	osName, err := getNameForOS(pc.OperatingSystem)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Invalid operating system specified %q, details = %v", pc.OperatingSystem, err),
+		}
+	}
+
+	teamID, err := strconv.Atoi(c.TeamID)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Invalid TeamID, details = %v", err),
+		}
+	}
+
+	var planID int
+	plans, _, err := client.Plans.List(teamID)
+	for _, plan := range plans {
+		if plan.Name == c.Plan {
+			planID = plan.ID
+			break
+		}
+	}
+
+	if c.Tags == nil {
+		c.Tags = map[string]string{}
+	}
+	c.Tags[machineUIDTagKey] = string(machine.UID)
+
+	serverCreateRequest := cherrygo.CreateServer{
```

Comment:
What happens when you create a server without an SSH key attached?

We provide the SSH key over the cloud-config, but some providers are sending the root password over the email if the SSH key isn't attached to the server on creation time.

Is this the case for CherryServers too? If it is the case, we should generate a random SSH key and attach it to server (there are helper functions for that, see DigitalOcean and Hetzner for examples).

Reply:
If you don't add ssh key a random root password is generated but not sent with email. It stays visible on the client dashboard for 24h then disappears.

Reply:
Are we good here ?

Reply:
I think we are for now

Reply:
Do you have an idea does including an SSH key prevents from receiving an email for each instance created? Right now we receive an email for each instance with some instance details, like name, size, etc, but that seems quite a bit spammy and we would like to avoid that.